### PR TITLE
Fix executable dictionary source hang

### LINKED ISSUE #1697: Executable external dictionary doesn't get any STDIN
According to docs when layout of executable dict configured as complex_key_cache list of needed keys will be written in STDIN of program, but it doesn't.

I asked about my issue in [Telegram](https://t.me/clickhouse_ru) and @proller recommended me to try `<command>tee id.log</command>` to log STDIN. As result  I got empty log.

Checked on Clickhouse v1.1.54318, v1.1.54327.
OS: Centos 7 (built using Altinity's script)

### LINKED ISSUE #2455: Выполнение linux команды через словари
Добрый день. Подскажите пожалуйста как работает блок execute command в словарях?

Мне нужно выполнять линукс команду, которая будет заменять все html entities на нужные символы.
Например у меня в таблице есть колонка name - в которой есть сущности &quot; &apos; 

Мне нужно заменить в имени каждой строки эти сущности на правильный знак
с помощью команды:

`echo 'Rib &lt;' | recode html..ascii `

Можно ли это сделать с помощью словарей?

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix executable dictionary source hang. In previous versions, when using some formats (e.g. `JSONEachRow`) data was not feed to a child process before it outputs at least something. This closes #1697. This closes #2455.

### CODE REVIEW

**Review by nikitamikhaylov — APPROVED** [review]
I've read the code of ExecutableDictionarySource and everything near with it. Why do you think that BOM caused deadlock? I think it is because of std::packaged_task which is executed asynchronously (by itself, but we call it from another thread for unknown reason). So, we called operator() [here](https://github.com/ClickHouse/ClickHouse/blob/4bd5524da9d830b68fb16a2091e2de52ff748c7d/src/Dictionaries/ExecutableDictionarySource.cpp#L112), but according to [notes here](https://en.cppreference.com/w/cpp/thread/packaged_task/operator()) it is synchronized only with calling of `get()` on its std::future. I think it means that we can't guarantee whether or not the function is executed after the operator() is called. It seems that in our case the underlying function wasn't executed. The author wrote in the [issue](https://github.com/ClickHouse/ClickHouse/issues/1697#issuecomment-372563928), that ClickHouse writes to stdin only after the process writes something to stdout, and this behavior is explained by reasoning above.

File: tests/queries/skip_list.json [hunk 1]
```diff
@@ -18,7 +18,8 @@
         "00152_insert_different_granularity",
         "00151_replace_partition_with_different_granularity",
         "00157_cache_dictionary",
-        "01193_metadata_loading"
+        "01193_metadata_loading",
+        "01474_executable_dictionary" /// informational stderr from sanitizer at start
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply tests/queries/skip_list.json:22]: We do.

**Comment by alexey-milovidov** [comment]
Integration test:
`test_adaptive_granularity/test.py::test_version_update_two_nodes`
- flaky test @alesapin

**Comment by alexey-milovidov** [comment]
AST Fuzzer: https://github.com/ClickHouse/ClickHouse/issues/14677

**Pull request merged**